lib/entry/crossDCsStreamPersistence.js

```diff
@@ -0,0 +1,275 @@
+const endListPendingEntriesFileName = "endListPending.json";
+const stoppedEntriesPath = config.get('stoppedEntriesPath');
+const kalturaTypes = require('./../kaltura-client-lib/KalturaTypes');
+
+// todo:
+// consult with Guy whether to keep playlist and just update on disk or read/write...
+// I think its better to keep cache of playlist and just write to disk each update.
+
+class CrossDCsStreamPersistence {
+
+    constructor() {
+
+        this._presentationEndTimeRegex = /("presentationEndTime":\s*)(\d+)/;
+        this._expireTimeRegex = /("expirationTime":\s*)(\d+)/;
+        this.stoppedEntriesFileName = path.join(stoppedEntriesPath, endListPendingEntriesFileName);
+        // note: default value of streamPersistenceTimer must be less than lowest play window size.
+        // assuming min play window size is 30 secs, the default persistenceCheckIntervalMillisec
+        // is 20 secs. In addition there is a time budget used as margin to make sure that even
+        // in very loaded system race condition chances will be very low.
+        this._persistenceCheckIntervalMillisec = config.get('persistenceCheckIntervalMillisec');
+        this._entriesToProcess = {};
+        this._playlistsData = {};
+        this._lastProcessedTimestampe = Date.now();
+        // _timeBudget purpose is to avoid race between packager and liveController, in case
+        // presentationEndTime value becomes past time before it is update by persistence loop.
+        this._timeBudget = 5000; // arbitrary time budget
+    }
+
+    streamPersistenceCheck(entryId, newEntry = false) {
+
+        let entryIsLive = false;
+        let ended = false;
+        let now = Date.now();
+        let nowStr = new Date(now).toLocaleString();
+        let presentationEndTime = now + 7200000;
+        let presentationTimeStr = new Date(presentationEndTime).toLocaleString();
+        
+        logger.debug(`====> [streamPersistenceCheck] starting stream persistence check`);
+        // todo: consult with Guy whether to filter received nodes objects according to status...
+        //entryServerNodeFilter.statusEqual = KalturaLiveStatus.PLAYABLE or KalturaLiveStatus.BROADCASTING;
+        return backendClient.getLiveEntryServerNodes(entryId)
+            .then((serverObjs) => {
+                return _.reduce(serverObjs, (memo, dc) => {
+                    return (dc.status === kalturaTypes.KalturaEntryServerNodeStatus.PLAYABLE) || memo;
+                }, false);
+            })
+            .then((isLive) => {
+                entryIsLive = isLive;
+                return qio.read(this._playlistsData[entryId].path);
+            })
+            .then((content) => {
+                let playlist = content;
+                this._playlistsData[entryId].lastProcessedTime = now;
+                if (newEntry) {
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] STOPPED STREAMING expiration time set to [${nowStr}]`);
```

Comment:
the logger should be consist with all our log (entry id in the begining)
look at the log file